src/utils.js

```diff
@@ -86,3 +86,79 @@ export const openIOSARQuickLook = url => {
   anchor.appendChild(document.createElement('img'));
   anchor.click();
 };
+
+
+/**
+ * This helper analyzes the layout of the current page to decide if we should
+ * use the natural device pixel ratio, or a capped value.
+ *
+ * We cap DPR if there is no meta viewport (suggesting that user is not
+ * consciously specifying how to scale the viewport relative to the device
+ * screen size) and also the window dimensions are a multiple of the effective
+ * screen dimensions (in CSS pixels).
+ *
+ * The rationale for this is that these conditions typically lead to a
+ * pathological outcome on mobile devices. When the window dimensions are
+ * scaled up on a device with a high DPR, we create a canvas that is much
+ * larger than appropriate to accomodate for the pixel density.
+ *
+ * This value needs to be measured in real time, as device pixel ratio can
+ * change over time (e.g., when a user zooms the page). Also, in some cases
+ * (such as Firefox on Android), the window's innerWidth is initially reported
+ * as the same as the screen's availWidth but changes later.
+ *
+ * A user who specifies a meta viewport, thereby consciously creating scaling
+ * conditions where <model-viewer> is slow, will be encouraged to live their
+ * best life.
+ */
+export const resolveDpr = (() => {
+  // The ratio we use for a "capped" scenario:
+  const CAPPED_DEVICE_PIXEL_RATIO = 1;
+
+  // If true, implies that the user is conscious of the viewport scaling
+  // relative to the device screen size.
+  const HAS_META_VIEWPORT_TAG = (() => {
+    const metas = document.head != null ?
+        Array.from(document.head.querySelectorAll('meta')) :
+        [];
+
+    for (const meta of metas) {
+      if (meta.name === 'viewport') {
+        return true;
+      }
+    }
+
+    return false;
+  })();
+
+  // NOTE(cdata): f true, this suggests that the window contents have been
+  // scaled such that the inner width is a multiple of the effective CSS pixel
+  // dimensions of the actual screen. This is assumed to imply one of the
+  // following conditions:
+  //
+  //  1. The window has been manually sized to extent beyond the bounds of
+  // the user's display (probably only possible in desktop scenarios).
+  //  2. There is no meta viewport on a high-DPR device, or else the meta
+  // viewport has scaled the defuault width higher than the screen width of
+  // the device.
+  //
+  // Condition #1 is not very common, but certainly possible. As long as there
+  // is a meta viewport, we won't cap the DPR.
+  // Condition #2 is usually not intentional but very easy to arrive at (a
+  // user need only omit the meta viewport tag and view their site on a
+  // contemporary mobile device).
+  //
+  // @see https://www.quirksmode.org/dom/w3c_cssom.html#screenview
+  const hasRiskyWindowToScreenRatio = () =>
+      Math.max(
+          window.innerWidth / window.screen.availWidth,
+          window.innerHeight / window.screen.availHeight) > 1;
```

Comment:
It's been awhile since I've used `availWidth|Height`; found this note on MDN:

"Internet Explorer will take into account the zoom setting when reporting the screen width. It will only return the real width of the screen if the zoom is set to 100%."

Just a comment, not sure if this is addressed

Reply:
Edit: looks like quirksmode doesn't address IE10+, but unclear what IE the [MDN notes for screen.width](https://developer.mozilla.org/en-US/docs/Web/API/Screen/width#Notes) refers to